File: apps/backend/app/schemas/analytics.py
from typing import Optional, Dict, Any, List
from pydantic import BaseModel, ConfigDict
from datetime import datetime

# Common Sub-Schemas
class Insight(BaseModel):
    id: str
    title: str
    description: str
    severity: str # "Info", "Warning", "Critical"
    confidence: str # "High", "Medium", "Low"
    category: str # "Correlation", "Distribution", "Trend", "Outlier", "Data Quality"
    affected_columns: List[str] = []
    recommendation: Optional[str] = None

class FeatureImportanceItem(BaseModel):
    column: str
    overall_importance_score: float
    variance_score: float
    correlation_score: float
    missing_value_penalty: float
    cardinality_score: float
    information_density: float

# Main Analytics Payload Schemas
class CorrelationResult(BaseModel):
    pearson_matrix: Dict[str, Dict[str, float]]
    spearman_matrix: Dict[str, Dict[str, float]]
    covariance_matrix: Dict[str, Dict[str, float]]
    strong_correlations: List[Dict[str, Any]] # e.g. [{"col1": "A", "col2": "B", "score": 0.95, "type": "Pearson"}]

class DistributionResult(BaseModel):
    columns: Dict[str, Dict[str, Any]] # e.g. "Age": {"mean": 30, "median": 29, "classification": "Normal", ...}

class TrendResult(BaseModel):
    increasing_trends: List[Dict[str, Any]]
    decreasing_trends: List[Dict[str, Any]]
    rolling_averages: Dict[str, Dict[str, Any]]
    seasonality_indicators: Optional[Dict[str, Any]] = None

class OutlierResult(BaseModel):
    outlier_count: int
    outlier_percentage: float
    affected_columns: List[str]
    extreme_values: Dict[str, List[Dict[str, Any]]] # column -> list of extreme values/details
    data_quality_warnings: List[str]

class FeatureImportanceResult(BaseModel):
    ranked_features: List[FeatureImportanceItem]

class ExecutiveSummary(BaseModel):
    dataset_quality_score: int # 0-100
    readiness_score: int # 0-100
    completeness: float # percentage
    strongest_findings: List[Insight]
    weakest_areas: List[Insight]
    potential_risks: List[Insight]
    executive_summary_text: str

# Response Schema for API
class AnalyticsResultResponse(BaseModel):
    id: str
    dataset_id: str
    analysis_version: Optional[str] = None
    
    correlation_data: Optional[CorrelationResult] = None
    distribution_data: Optional[DistributionResult] = None
    trend_data: Optional[TrendResult] = None
    outlier_data: Optional[OutlierResult] = None
    feature_importance: Optional[FeatureImportanceResult] = None
    insights: Optional[List[Insight]] = None
    executive_summary: Optional[ExecutiveSummary] = None
    
    processing_duration: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: apps/frontend/src/features/analytics/types/analytics.ts
export interface Insight {
  id: string;
  title: string;
  description: string;
  severity: "Info" | "Warning" | "Critical";
  confidence: "High" | "Medium" | "Low";
  category: string;
  affected_columns: string[];
  recommendation?: string;
}

export interface FeatureImportanceItem {
  column: string;
  overall_importance_score: number;
  variance_score: number;
  correlation_score: number;
  missing_value_penalty: number;
  cardinality_score: number;
  information_density: number;
}

export interface CorrelationResult {
  pearson_matrix: Record<string, Record<string, number>>;
  spearman_matrix: Record<string, Record<string, number>>;
  covariance_matrix: Record<string, Record<string, number>>;
  strong_correlations: Array<{
    col1: string;
    col2: string;
    score: number;
    method: string;
    strength: string;
    direction: string;
  }>;
}

export interface DistributionResult {
  columns: Record<string, any>; // Complex nested object
}

export interface TrendResult {
  increasing_trends: Array<{
    column: string;
    time_column: string;
    growth_percentage: number;
    slope: number;
  }>;
  decreasing_trends: Array<{
    column: string;
    time_column: string;
    growth_percentage: number;
    slope: number;
  }>;
  rolling_averages: Record<string, { dates: string[]; values: number[] }>;
  seasonality_indicators?: Record<string, any>;
}

export interface OutlierResult {
  outlier_count: number;
  outlier_percentage: number;
  affected_columns: string[];
  extreme_values: Record<
    string,
    Array<{ index: number; value: number; type: string }>
  >;
  data_quality_warnings: string[];
}

export interface FeatureImportanceResult {
  ranked_features: FeatureImportanceItem[];
}

export interface ExecutiveSummary {
  dataset_quality_score: number;
  readiness_score: number;
  completeness: number;
  strongest_findings: Insight[];
  weakest_areas: Insight[];
  potential_risks: Insight[];
  executive_summary_text: string;
}

export interface AnalyticsResultResponse {
  id: string;
  dataset_id: string;
  analysis_version?: string;
  correlation_data?: CorrelationResult;
  distribution_data?: DistributionResult;
  trend_data?: TrendResult;
  outlier_data?: OutlierResult;
  feature_importance?: FeatureImportanceResult;
  insights?: Insight[];
  executive_summary?: ExecutiveSummary;
  processing_duration?: number;
  created_at: string;
  updated_at: string;
}
